### ISSUE TITLE
Trying to import Hybrid connection from different resource group

### ISSUE THREAD

**Issue by angeloCabezas** [hashicorp/terraform-provider-azurerm#14845]
Hi, 

I've been trying to import "azurerm_app_service_hybrid_connection" but I get the next exception :

" unable to List Access Keys for Namespace "company-pp-dev-hc" (Resource Group "Company_PP_Staging"): namespaces.NamespacesClient#ListKeys: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: error response cannot be parsed: "\"The requested resource defaultSender does not exist.\"" error: json: cannot unmarshal string into Go value of type azure.RequestError"

Scenary :

- My app services is in the resource group "Company_PP_Dev"
- My hybrid connections are in a different resource group, basically the relay is in the group "Company_PP_Staging"

my import command is like this : 

azurerm_app_service_hybrid_connection.runtimeGroup_designer_hybrid_connection_splunkapicompanycom_xxxx/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/Company_PP_Staging/providers/Microsoft.Web/sites/company-pp-staging/hybridConnectionNamespaces/company-pp-dev-hc/relays/nameHybridConnection

my guessing is that relay namespace is in another resource group.

Could you pls help me to solve this issue? or give me an alternative to do that? (btw I've already been looking for a solution for some time). 

Many thanks.

**Comment by lonegunmanb** [comment]
Hi @angeloCabezas , would please give a minimum sample code and procedures that can reproduce this issue? I'd like to dig in but I don't know the exactly configuration you've used, so a little more information would be a great help, thanks!

**Comment by angeloCabezas** [comment]
Hi @lonegunmanb, yes sure.

Issue : We are not able to import "azurerm_app_service_hybrid_connection" resources.

Steps to replicate : 

1 - Have an App Service in resource Group ("A") (My case : Cignium_LM_Staging)
2 - Have an HybridConnection in resource Group ("B") (My case : Cignium_LM_Dev)
3 - Try to import a relationship between App Service and HybridConnection by terraform import command("azurerm_app_service_hybrid_connection")
See what we get in the next image:
![ImportProblem](https://user-images.githubusercontent.com/36689256/148575175-34a5c3e9-7098-4007-8752-1d9d054998aa.jpg)

Finally, this is what we have in **cdk.tf.json**, but this is relative, the most important part is to have the object created with this identifier (runtimeGroup_designer_hybrid_connection_splunkapicigniumcom_F68E6D21) to match the terraform import.

`"azurerm_app_service_hybrid_connection": {
      "runtimeGroup_designer_hybrid_connection_splunkapicigniumcom_F68E6D21": {
        "app_service_name": "${azurerm_app_service.runtimeGroup_appService_6A2D8FA7.name}",
        "hostname": "xxxx.xxxx.com",
        "port": 8088,
        "relay_id": "${azurerm_relay_hybrid_connection.splunk.id}",
        "resource_group_name": "${azurerm_relay_hybrid_connection.splunk.resource_group_name}",
        "//": {
          "metadata": {
            "path": "lm/runtimeGroup/designer_hybrid_connection_xxxx.xxx.com",
            "uniqueId": "runtimeGroup_designer_hybrid_connection_splunkapicigniumcom_F68E6D21"
          }
        }
      }
    }`

Let me know any new, many thanks.

**Comment by lonegunmanb** [comment]
Hi @angeloCabezas , I have a question.

In your import command we can see the id you want import contains `/subscriptions/xxx/resourceGroups/Cignium_LM_Staging/providers/...`, meanwhile in your `azurerm_app_service_hybrid_connection` definition the `resource_group_name` is `"${azurerm_relay_hybrid_connection.splunk.resource_group_name}"`, and as you said:

>Have an HybridConnection in resource Group ("B") (My case : Cignium_LM_Dev)

I think the `azurerm_app_service_hybrid_connection.resource_group_name` should be `Cignium_LM_Dev` instead of `Cignium_LM_Staging` right?

And as you want import this existing `azurerm_app_service_hybrid_connection`, may I ask how you created this connection at first place so I can try to reproduce it? Thanks!

**Comment by angeloCabezas** [comment]
Hi @lonegunmanb 

I think the `azurerm_app_service_hybrid_connection.resource_group_name` should be `Cignium_LM_Dev` instead of `Cignium_LM_Staging` right?

Not really, Because the app service is in `Cignium_LM_Staging` so this relationship should be here in the same way. (Also if I put Cignium_LM_Dev, it said "Resource not found" because terraform import try to find de app service in Cignium_LM_Dev and that's not right)

Yes, right panel in the app service, it is a section calls "NetWorking" choose that and then select hybrid connection. So in that point you need to chose an hybrid connection for a relation.

<img width="855" alt="image" src="https://user-images.githubusercontent.com/36689256/148979667-eb552ec6-b61d-4119-aa3a-e825309c2f0e.png">

Let me know any question. thanks

**Comment by lonegunmanb** [comment]
Hello @angeloCabezas , thanks for your reply, I've reproduced this issue, and after digged in the code, I've found that the provider DO support different resource groups, but in a different way.

in project's `{PROJECT_ROOT}\internal\services\web\app_service_hybrid_connection_resource_test.go`, there's a test showed how we should provision connection in res group different than app servcie's res group: 

```go
func TestAccAppServiceHybridConnection_differentResourceGroup(t *testing.T) {
	data := acceptance.BuildTestData(t, "azurerm_app_service_hybrid_connection", "test")
	r := AppServiceHybridConnectionResource{}

	data.ResourceTest(t, r, []acceptance.TestStep{
		{
			Config: r.differentResourceGroup(data),
			Check: acceptance.ComposeTestCheckFunc(
				check.That(data.ResourceName).ExistsInAzure(r),
			),
		},
		data.ImportStep(),
	})
}
```

Where `r.differentResourceGroup` method is:

```go
func (r AppServiceHybridConnectionResource) differentResourceGroup(data acceptance.TestData) string {
	return fmt.Sprintf(`
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-%d"
  location = "%s"
}

resource "azurerm_resource_group" "relay" {
  name     = "acctestRG-relay-%d"
  location = "%s"
}

resource "azurerm_app_service_plan" "test" {
  name                = "acctest-ASP-%d"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name

  sku {
    tier = "Standard"
    size = "S1"
  }
}

resource "azurerm_app_service" "test" {
  name                = "acctest-AS-%d"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  app_service_plan_id = azurerm_app_service_plan.test.id
}

resource "azurerm_relay_namespace" "test" {
  name                = "acctest-RN-%d"
  location            = azurerm_resource_group.relay.location
  resource_group_name = azurerm_resource_group.relay.name

  sku_name = "Standard"
}

resource "azurerm_relay_hybrid_connection" "test" {
  name                 = "acctest-RHC-%d"
  resource_group_name  = azurerm_resource_group.relay.name
  relay_namespace_name = azurerm_relay_namespace.test.name
  user_metadata        = "metadatatest"
}

resource "azurerm_app_service_hybrid_connection" "test" {
  app_service_name    = azurerm_app_service.test.name
  resource_group_name = azurerm_resource_group.test.name
  relay_id            = azurerm_relay_hybrid_connection.test.id
  hostname            = "testhostname.azuretest"
  port                = 443
  send_key_name       = "RootManageSharedAccessKey"
}
`, data.RandomInteger, data.Locations.Primary, data.RandomInteger, data.Locations.Primary, data.RandomInteger, data.RandomInteger, data.RandomInteger, data.RandomInteger)
}
```

Let me extract tf code out of go code:

```hcl
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-%d"
  location = "%s"
}

resource "azurerm_resource_group" "relay" {
  name     = "acctestRG-relay-%d"
  location = "%s"
}

resource "azurerm_app_service_plan" "test" {
  name                = "acctest-ASP-%d"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name

  sku {
    tier = "Standard"
    size = "S1"
  }
}

resource "azurerm_app_service" "test" {
  name                = "acctest-AS-%d"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  app_service_plan_id = azurerm_app_service_plan.test.id
}

resource "azurerm_relay_namespace" "test" {
  name                = "acctest-RN-%d"
  location            = azurerm_resource_group.relay.location
  resource_group_name = azurerm_resource_group.relay.name

  sku_name = "Standard"
}

resource "azurerm_relay_hybrid_connection" "test" {
  name                 = "acctest-RHC-%d"
  resource_group_name  = azurerm_resource_group.relay.name
  relay_namespace_name = azurerm_relay_namespace.test.name
  user_metadata        = "metadatatest"
}

resource "azurerm_app_service_hybrid_connection" "test" {
  app_service_name    = azurerm_app_service.test.name
  resource_group_name = azurerm_resource_group.test.name
  relay_id            = azurerm_relay_hybrid_connection.test.id
  hostname            = "testhostname.azuretest"
  port                = 443
  send_key_name       = "RootManageSharedAccessKey"
}
```

As we can see, it should works, if we put out `azurerm_app_service_hybrid_connection` in the same res group that we put our `azurerm_relay_namespace`. The error you've seen is triggered by relay namespace client retrieving namespace's primary access key. The test code passed the test:

=== RUN   TestAccAppServiceHybridConnection_differentResourceGroup
=== PAUSE TestAccAppServiceHybridConnection_differentResourceGroup
=== CONT  TestAccAppServiceHybridConnection_differentResourceGroup
--- PASS: TestAccAppServiceHybridConnection_differentResourceGroup (340.80s)

I guess that in your code, the relay namespace's res group is different than `azurerm_app_service_hybrid_connection`'s res group right?

**Comment by angeloCabezas** [comment]
Hi @lonegunmanb , thank you so much for your time.

I guess that in your code, the relay namespace's res group is different than azurerm_app_service_hybrid_connection's res group right?
Yes, it is.

I know that you right in how to use other stuffs (hybrid connections, services bus ... etc) from different resources group by coding, the main problem is when I tried to import that resources from Azure.

As you can see in my first image when I try to import azurerm_app_service_hybrid_connection I get that error.

Could you try to import the resource to your terraform.tfstate, pls? 

Many thanks.!!

**Comment by lonegunmanb** [comment]
Hi @angeloCabezas , I'm following this issue. In your error message the service complain "The requested resource defaultSender does not exist."

The error is caused by listing access key based on `azurerm_app_service_hybrid_connection.send_key_name`, the service side returned 404. I can try a fix, but I'd like to gather more knowledge since I'm not familiar with hybrid connection and relay so please forgive me.

In `azurerm_app_service_hybrid_connection`'s [document](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/app_service_hybrid_connection#send_key_name), the `send_key_name` means:

>The name of the Service Bus key which has Send permissions. Defaults to `RootManageSharedAccessKey`.

I found that the `RootManageSharedAccessKey` was created by default on the relay, and based on the error message it seems like your existing `azurerm_app_service_hybrid_connection.send_key_name` is "defaultSender" right? Would you please check whether there is an access key name "defaultSender" on your relay or not? Thanks!

**Comment by angeloCabezas** [comment]
Hi @lonegunmanb 

No worries, I'm not familiar with hybrid connections too but I have to pass the information to terraform. 

I've been checking the information, so in order to answer your questions 

I found that the RootManageSharedAccessKey was created by default on the relay, and based on the error message it seems like your existing hybrid connection's send_key_name is "defaultSender" right? 

Yes, for this splunk hybrid connection I have two shared access policies (defaultListener , defaultSender), they were already imported.

Would you please check whether there is an access key name "defaultSender" on your relay's Shared access policies or not?

Yes, it is an access key name "DefaultSender" and "defaultListener"

if this "defaultSender" is absent, would this issue be fixed by adding a shared access policy in your relay?

No, this "defaultSender" is not basent, the issue still there.

This is the structure that I have

>       Relay : Cignium-lm-dev-hc (Shared Access Policy : RootManageSharedAccessKey )

>       Entity hybrid connection : splunk (Shared Access Policy : defaultListener , defaultSender)

Could you try to import the resources that you created in a previous comment ?? 

Looks like you have the right scenary, so most likely you will get the same error.

Many thanks.!!

**Comment by lonegunmanb** [comment]
Hi @angeloCabezas , thanks for your reply, I think I've figured out this issue. 

The source of this error is provider wants query `send_key_value` as attribute this resource export via listing key in relay by shared access policies on your hybrid connection. Your `defaultListener` and `defaultSender` were not on the relay, so an error was raised complained that. 

May I ask how you add these two shared access policies to your hybrid connection? How should we get this service bus access key's value since we cannot find it on the relay? Thanks!

**Comment by lonegunmanb** [comment]
Hi @angeloCabezas , this issue was caused by the reason I've mentioned, we don't need to import the resource, creating a resource with shared access policies that relay doesn't have will do the same. I need to figure out how to get the `send_key_value` in this case since the old way that queries it from the relay doesn't work.

So what you have provided is the key point. I'll try your way and try to query from the hybrid connection if no such policy on the relay.

And thank you for your patience and detailed information, I think we are close to a fix.

**Comment by lonegunmanb** [comment]
Hi @angeloCabezas , I've submitted a patch to fix this issue, I believe it is irrelevant to two different resource groups. The patch will try to list access keys on the hybrid connection if the listing on the relay failed. Please give it a try when the pr has been merged and the new release has been published. Thanks for your patience.

**Comment by lonegunmanb** [comment]
Hi @angeloCabezas , I have a further question: are you sure this shared policy that was declared on the hybrid connection is the same concept as the one declared on the relay? According to the [document](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/app_service_hybrid_connection), `send_key_name` means:

> (Optional) The name of the Service Bus key which has Send permissions. Defaults to RootManageSharedAccessKey.

I'm not familiar with hybrid connection, is the key on hybrid connection as same as "Service Bus Key" here? So in your scenario you can use either the key on hybrid connection or relay, they both do the job right? Would you please provide me with some documents about that?

**Comment by lonegunmanb** [comment]
Hi @angeloCabezas , we're close to a fix but I have a further question since I'm not familiar with this app service hybrid connection service.

I've successfully created an `azurerm_app_service_hybrid_connection` with `send_key_name` pointed to a shared policy that was declared on the hybrid connection not the relay, I can see the send key name and value was embedded into "GATEWAY CONNECTION STRING" of the app service hybrid connection. So both shared policies that are declared on the relay and the hybrid connection work in this "GATEWAY CONNECTION STRING" right?

**Comment by angeloCabezas** [comment]
Hey @lonegunmanb, nop the policies are like this way : 

`  Relay : Cignium-lm-dev-hc (Shared Access Policy : RootManageSharedAccessKey )`
`  Entity hybrid connection : splunk (Shared Access Policy : defaultListener , defaultSender)`

So what we have in my environment is that the "Gateway connection string" is using `SharedAccessKeyName=defaultListener`, so I would say it is using the hybrid connection SharedAccessKey. 

I saw your pullrequest it's already merged, Should I download this new version ?? 

Thanks.

**Comment by lonegunmanb** [comment]
> I saw your pullrequest it's already merged, Should I download this new version ??

Hi @angeloCabezas , pr #15118 **HAS NOT** been merged yet. 

>So what we have in my environment is that the "Gateway connection string" is using `SharedAccessKeyName=defaultListener`

Please check whether the secret in your connection string is the same as it defined in the connection, if so, I think this patch will work.

**Comment by lonegunmanb** [comment]
> Hi @lonegunmanb 
> 
> Yes, this primary connection string :
> 
> <img width="883" alt="image" src="https://user-images.githubusercontent.com/36689256/151830053-5161d5a8-d202-42a5-92b0-728039a38f16.png">
> 
> is same to this one (Gateway ...) :
> 
> <img width="313" alt="image" src="https://user-images.githubusercontent.com/36689256/151830447-72110f58-0e10-45da-9ce8-2e5ccdbbf1ba.png">
> 
> So, This fix will solve the issue right ?
> 
> Many thanks.

I think so because I can create such hybrid connection via this patch, please wait until HashiCorp has finished the review and merge and publish a new release.

**Comment by angeloCabezas** [comment]
Hi @lonegunmanb 

Do you have an estimate date to use it? 

Many thanks.!!

**Comment by lonegunmanb** [comment]
> Hi @lonegunmanb 
> 
> Do you have an estimate date to use it? 
> 
> Many thanks.!!

Since the patch was released with v2.95.0, please try it with the latest version of plugin, thanks!
